src/postprocess/refine.jl

```diff
@@ -0,0 +1,183 @@
+# Refinement of some quantities of interest (density, forces) following the
+# strategy described in [CDKL2021].
+#
+# The overall strategy is to perform a first SCF,
+# transfer the result to a new basis set with a larger Ecut,
+# then perform an approximate first order correction in the larger space.
+#
+# To compute the first order correction ("refinement") δP in the larger space,
+# one needs to invert (Ω+K)δP = -R(P) where (Ω+K) is the Hessian matrix and
+# R(P) = [P, [P, H(P)]] is the residual.
+# See newton.jl for an explanation of these quantities.
+#
+# The basis elements of the smaller basis set are called low frequency (_1 suffix).
+# The additional basis elements are all high frequency (_2 suffix).
+# We approximate (Ω+K)_22 by M_22 which is cheap to compute and invert, and (Ω+K)_21 by 0.
+#
+# The metric operator M is as follows:
+# M applied to the i-th band is defined by
+#   M_i = P^⟂ T_i^{1/2} P^⟂ T_i^{1/2} P^⟂,
+# with the diagonal operator
+#   T_i = kinetic energy + mean kinetic energy of i-th band.
+# To invert M_i effectively, we use P^⟂ T_i^{-1} P^⟂ as the preconditioner.
+#
+# Solving the system then amounts to computing
+#   δP_2 = M^{-1}_22 R_2(P)
+#   δP_1 = (Ω+K)^{-1}_11 (R_1(P) - (Ω+K)_12 δP_2)
+# The inversion of (Ω+K)_11 is roughly as expensive as the SCF on the low frequency space.
+#
+# [CDKL2021]:
+#     E. Cancès, G. Dusson, G. Kemlin, and A. Levitt
+#     *Practical error bounds for properties in plane-wave electronic structure
+#     calculations* Preprint, 2021. [arXiv](https://arxiv.org/abs/2111.01470)
+
+"""
+Invert the metric operator M.
+"""
+function invert_refinement_metric(basis::PlaneWaveBasis{T}, ψ, res) where {T}
+    # Apply the M_i operator with i=n.
+    function apply_M(ψk, Pk, δψnk, n)
+        proj_tangent_kpt!(δψnk, ψk)
+        δψnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δψnk
+        proj_tangent_kpt!(δψnk, ψk)
+        δψnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δψnk
+        proj_tangent_kpt!(δψnk, ψk)
+    end
+
+    # Apply the M_i^{-1} operator with i=n.
+    function apply_inv_M(ψk, Pk, resk, n)
+        proj_tangent_kpt!(resk, ψk)
+        op(x) = apply_M(ψk, Pk, x, n)
+        function f_ldiv!(x, y)
+            x .= proj_tangent_kpt(y, ψk)
+            x ./= (Pk.mean_kin[n] .+ Pk.kin)
+            proj_tangent_kpt!(x, ψk)
+        end
+        J = LinearMap{eltype(ψk)}(op, size(resk, 1))
+        # This CG seems to converge very quickly in practice (often 1 iteration).
+        # Set a very low tolerance and warn if the iteration count is too high.
+        (δψk, history) = IterativeSolvers.cg(J, resk;
```

Comment:
Maybe not a bad idea to also use our internal CG.

Reply:
Done